src/vm/assembly.cpp

```diff
@@ -1790,6 +1790,20 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
         GCX_COOP();
 
         pMeth = GetEntryPoint();
+        {
+            GCX_PREEMP();
+
+#ifdef FEATURE_COMINTEROP
+            Thread::ApartmentState state = Thread::AS_Unknown;
+            state = SystemDomain::GetEntryPointThreadAptState(pMeth->GetMDImport(), pMeth->GetMemberDef());
+
+            // If the entry point has an explicit thread apartment state, set it
+            // before running the AppDomainManager initialization code.
+            if (state == Thread::AS_InSTA || state == Thread::AS_InMTA)
+                SystemDomain::SetThreadAptState(state);
+#endif // FEATURE_COMINTEROP
+        }
+
         if (pMeth) {
```

Comment:
Your new code should be undef this `if` to handle the case where `pMeth` is NULL.